File: netlify/functions/spellcheck.js
/**
 * Tibetan Spellchecker - Netlify Serverless Function
 * 基于 tibetan-spellchecker 词典资源
 */
const fs = require('fs');
const path = require('path');

const TSHEG = '\u0f0b';
const SHAD = '\u0f0d';

function loadDictionary(baseDir) {
  const valid = new Set();
  const syllablesDir = path.join(baseDir, 'syllables');

  const suffixes = JSON.parse(
    fs.readFileSync(path.join(syllablesDir, 'suffixes.json'), 'utf8')
  );
  suffixes.NB = [''];

  function loadRootTxt(filePath) {
    fs.readFileSync(filePath, 'utf8')
      .split('\n')
      .forEach((line) => {
        line = line.trim();
        if (!line || line.startsWith('#')) return;
        if (line.includes('/')) {
          const [root, suffixType] = line.split('/', 2).map((s) => s.trim());
          if (suffixes[suffixType]) {
            suffixes[suffixType].forEach((suf) => valid.add(root + suf));
          }
        }
      });
  }

  function loadWithSuffixes(filePath) {
    if (!fs.existsSync(filePath)) return;
    fs.readFileSync(filePath, 'utf8')
      .split('\n')
      .forEach((line) => {
        line = line.trim();
        if (!line) return;
        if (line.includes('/')) {
          const [root, suffixType] = line.split('/', 2).map((s) => s.trim());
          if (suffixes[suffixType]) {
            suffixes[suffixType].forEach((suf) => valid.add(root + suf));
          }
        } else {
          valid.add(line);
        }
      });
  }

  function loadPlain(filePath) {
    if (!fs.existsSync(filePath)) return;
    fs.readFileSync(filePath, 'utf8')
      .split('\n')
      .forEach((line) => {
        line = line.trim();
        if (line) valid.add(line);
      });
  }

  loadRootTxt(path.join(syllablesDir, 'root.txt'));
  loadWithSuffixes(path.join(syllablesDir, 'wasurs.txt'));
  loadWithSuffixes(path.join(syllablesDir, 'rare.txt'));
  loadPlain(path.join(syllablesDir, 'exceptions.txt'));
  loadWithSuffixes(path.join(syllablesDir, 'proper-names.txt'));

  // supplement.txt - 补充词典
  const supplementPath = path.join(syllablesDir, 'supplement.txt');
  if (fs.existsSync(supplementPath)) {
    fs.readFileSync(supplementPath, 'utf8')
      .split('\n')
      .forEach((line) => {
        line = line.trim();
        if (line && !line.startsWith('#')) valid.add(line);
      });
  }

  return valid;
}

function spellcheckText(text, validSyllables) {
  const result = [];
  let start = 0;

  for (let i = 0; i <= text.length; i++) {
    if (i === text.length || text[i] === TSHEG) {
      if (i > start) {
        const syl = text.slice(start, i);
        if (syl && !/^\s+$/.test(syl)) {
          const hasTibetan = /[\u0f00-\u0fff]/.test(syl);
          if (hasTibetan) {
            const lookup = syl.replace(new RegExp(SHAD + '+$'), '');
            result.push({
              syllable: syl,
              start,
              end: i,
              valid: validSyllables.has(lookup),
            });
          }
        }
      }
      start = i + 1;
    }
  }
  if (start < text.length) {
    const syl = text.slice(start);
    if (syl && !/^\s+$/.test(syl) && /[\u0f00-\u0fff]/.test(syl)) {
      const lookup = syl.replace(new RegExp(SHAD + '+$'), '');
      result.push({
        syllable: syl,
        start,
        end: text.length,
        valid: validSyllables.has(lookup),
      });
    }
  }

  return result;
}

// 词典在首次调用时加载并缓存
let validSyllables = null;

function getDictionary() {
  if (!validSyllables) {
    // Netlify 打包后 included_files 会在项目根目录，使用 process.cwd()
    const baseDir = process.cwd();
    validSyllables = loadDictionary(baseDir);
  }
  return validSyllables;
}

exports.handler = async (event, context) => {
  const headers = {
    'Content-Type': 'application/json',
    'Access-Control-Allow-Origin': '*',
    'Access-Control-Allow-Headers': 'Content-Type',
  };

  if (event.httpMethod === 'OPTIONS') {
    return { statusCode: 204, headers, body: '' };
  }

  if (event.httpMethod !== 'POST') {
    return {
      statusCode: 405,
      headers,
      body: JSON.stringify({ detail: 'Method not allowed' }),
    };
  }

  try {
    const body = JSON.parse(event.body || '{}');
    const text = body.text || '';

    const dict = getDictionary();
    const results = spellcheckText(text, dict);
    const errorCount = results.filter((r) => !r.valid).length;

    return {
      statusCode: 200,
      headers,
      body: JSON.stringify({
        results,
        text,
        error_count: errorCount,
      }),
    };
  } catch (err) {
    return {
      statusCode: 500,
      headers,
      body: JSON.stringify({ detail: err.message }),
    };
  }
};

File: spellcheck.py
"""
Tibetan Syllable Spellchecker
Uses dictionary resources from tibetan-spellchecker project
"""
import json
from pathlib import Path

# Tibetan tsheg (syllable separator) U+0F0B
TSHEG = "\u0f0b"
# Tibetan shad (period) U+0F0D - often appears at end of syllable
SHAD = "\u0f0d"


def load_dictionary(base_dir: Path) -> set:
    """Load all valid syllables from dictionary files."""
    valid = set()
    syllables_dir = base_dir / "syllables"

    # Load suffixes
    with open(syllables_dir / "suffixes.json", "r", encoding="utf-8") as f:
        suffixes = json.load(f)

    # Add NB (no suffix) type - only empty suffix
    suffixes["NB"] = [""]

    # Load root.txt: format is "root/TYPE" (e.g., "ཀ/A", "དཀ/NB")
    with open(syllables_dir / "root.txt", "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue
            if "/" in line:
                root, suffix_type = line.split("/", 1)
                root = root.strip()
                suffix_type = suffix_type.strip()
                if suffix_type in suffixes:
                    for suf in suffixes[suffix_type]:
                        valid.add(root + suf)

    # Load wasurs.txt: format "syllable" or "syllable/TYPE" (full form or root+type)
    with open(syllables_dir / "wasurs.txt", "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            if "/" in line:
                root, suffix_type = line.split("/", 1)
                if suffix_type.strip() in suffixes:
                    for suf in suffixes[suffix_type.strip()]:
                        valid.add(root.strip() + suf)
            else:
                valid.add(line)

    # Load rare.txt: full syllable forms, some with /C
    with open(syllables_dir / "rare.txt", "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            if "/" in line:
                root, suffix_type = line.split("/", 1)
                if suffix_type.strip() in suffixes:
                    for suf in suffixes[suffix_type.strip()]:
                        valid.add(root.strip() + suf)
            else:
                valid.add(line)

    # Load exceptions.txt: full forms
    with open(syllables_dir / "exceptions.txt", "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if line:
                valid.add(line)

    # Load proper-names.txt
    with open(syllables_dir / "proper-names.txt", "r", encoding="utf-8") as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            if "/" in line:
                root, suffix_type = line.split("/", 1)
                if suffix_type.strip() in suffixes:
                    for suf in suffixes[suffix_type.strip()]:
                        valid.add(root.strip() + suf)
            else:
                valid.add(line)

    # Load supplement.txt - 补充词典（常用但未生成的音节，用户可编辑添加）
    supplement_file = syllables_dir / "supplement.txt"
    if supplement_file.exists():
        with open(supplement_file, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith("#"):
                    valid.add(line)

    return valid


def split_syllables(text: str) -> list[tuple[str, int, int]]:
    """
    Split Tibetan text into syllables by tsheg.
    Returns list of (syllable, start_idx, end_idx) for each syllable.
    """
    result = []
    start = 0
    for i, char in enumerate(text):
        if char == TSHEG:
            if i > start:
                result.append((text[start:i], start, i))
            start = i + 1
    if start < len(text):
        result.append((text[start:], start, len(text)))
    return result


def spellcheck_text(text: str, valid_syllables: set) -> list[dict]:
    """
    Check Tibetan text for spelling errors.
    Returns list of {syllable, start, end, valid} for each syllable.
    """
    syllables = split_syllables(text)
    results = []
    for syl, start, end in syllables:
        # Skip empty, spaces, punctuation
        if not syl or syl.isspace():
            continue
        # Skip non-Tibetan (numbers, etc.)
        if not any(ord(c) >= 0x0F00 and ord(c) <= 0x0FFF for c in syl):
            continue
        # Strip trailing shad for dictionary lookup (ལེགས། -> ལེགས)
        lookup = syl.rstrip(SHAD)
        results.append({
            "syllable": syl,
            "start": start,
            "end": end,
            "valid": lookup in valid_syllables,
        })
    return results
